samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -461,8 +482,25 @@ public void onJobModelExpired() {
             } else {
               LOGGER.info("Container: {} shutdown completed for stream processor: {}.", container, processorId);
             }
+          } else if (state == State.IN_REBALANCE) {
+            if (container != null) {
+              boolean hasContainerShutdown = interruptContainerAndShutdownExecutorService();
+              if (!hasContainerShutdown) {
+                LOGGER.warn("Job model expire unsuccessful. Failed to interrupt container: {} safely. "
+                    + "Stopping the stream processor: {}", container, processorId);
+                state = State.STOPPING;
+                jobCoordinator.stop();
+              } else {
+                Preconditions.checkState(containerExecutorService.isShutdown(),
```

Comment:
Good point. 
1. My intent was to ensure the previous executor service was terminated so that we don't leak executor service. But looking at the code deeper, i realized I can't use `isTerminated` because the container thread will contend for the lock to check the state both for `afterStop` and `afterFailure`. One option is to push the creation of executor service to OnNewJobModel to give executor service a chance to acquire the lock and finish the `afterX`. What do you think?
2. Exception will cause the job coordinator to shutdown and subsequently the stream processor as well.

Reply:
It seems a little unintuitive to try to do the clean-up in `onNewJobModel`.
Do you need to clean up the executor service given that the STARTED/RUNNING flow doesn't do it?
If you need to clean up the executor service, then do you actually need to hold the lock while it is waiting to shut down? That said, it's also complicated to try to do more granular locking, so I don't know what the ideal solution would be.

Reply:
Agreed. I left the executor service to be garbage collected for now so that I don't introduce more complexity to the code. Let me rethink about cleaning up the executor service separately and handle all the places (started/running) at once.